bdns/dns.go

```diff
@@ -287,25 +288,61 @@ func isPrivateV4(ip net.IP) bool {
 // provided hostname. This method assumes that the external resolver will chase
 // CNAME/DNAME aliases and return relevant A records.  It will retry requests in
 // the case of temporary network errors. It can return net package,
-// context.Canceled, and context.DeadlineExceeded errors.
+// context.Canceled, and context.DeadlineExceeded errors, all wrapped in the DNSError type.
 func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
-	var addrs []net.IP
-	dnsType := dns.TypeA
-	r, err := dnsResolver.exchangeOne(ctx, hostname, dnsType, dnsResolver.aStats)
-	if err != nil {
-		return addrs, &DNSError{dnsType, hostname, err, -1}
-	}
-	if r.Rcode != dns.RcodeSuccess {
-		return nil, &DNSError{dnsType, hostname, nil, r.Rcode}
+	errCh := make(chan error, 2)
+	var respA, respAAAA *dns.Msg
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		var err error
+		respA, err := dnsResolver.exchangeOne(ctx, hostname, dns.TypeA, dnsResolver.aStats)
```

Comment:
True, a better solution using the pattern I described bellow is replacing the error `chan` and pair of response vars with a single
```
chan struct{
  err error
  records []dns.RR
}
```
and unwrap the results in the `numRacers` for loop into a slice of records. This would also allow you to consolidate the two for loops you have for testing A/AAA validity into a single loop with a `switch typed := rr.(type)`.